extensions/catalog/federated-catalog-cache/src/main/java/org/eclipse/dataspaceconnector/catalog/cache/ExecutionManager.java

```diff
@@ -0,0 +1,240 @@
+/*
+ *  Copyright (c) 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.catalog.cache;
+
+import org.eclipse.dataspaceconnector.catalog.cache.crawler.CatalogCrawler;
+import org.eclipse.dataspaceconnector.catalog.spi.CrawlerErrorHandler;
+import org.eclipse.dataspaceconnector.catalog.spi.CrawlerSuccessHandler;
+import org.eclipse.dataspaceconnector.catalog.spi.FederatedCacheNodeDirectory;
+import org.eclipse.dataspaceconnector.catalog.spi.NodeQueryAdapterRegistry;
+import org.eclipse.dataspaceconnector.catalog.spi.WorkItem;
+import org.eclipse.dataspaceconnector.catalog.spi.model.ExecutionPlan;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.lang.String.format;
+
+public class ExecutionManager {
+
+    private Monitor monitor;
+    private Runnable preExecutionTask;
+    private Runnable postExecutionTask;
+    private FederatedCacheNodeDirectory directory;
+    private String connectorId;
+    private int numCrawlers = 1;
+    private NodeQueryAdapterRegistry nodeQueryAdapterRegistry;
+    private CrawlerSuccessHandler successHandler;
+
+    private ExecutionManager() {
+    }
+
+    public void executePlan(ExecutionPlan plan) {
+        plan.run(() -> {
+
+            monitor.debug(message("Run pre-execution task"));
+            runPreExecution();
+
+            // load work items from directory
+            List<WorkItem> workItems = fetchWorkItems();
+            if (workItems.isEmpty()) {
+                monitor.warning("No WorkItems found, aborting execution");
+                return;
+            }
+            monitor.debug(message("Loaded " + workItems.size() + " work items from storage"));
+            var allItems = new ArrayBlockingQueue<>(workItems.size(), true, workItems);
+
+            monitor.debug(message("Instantiate crawlers..."));
+            //instantiate fixed pool of crawlers
+            var errorHandler = createErrorHandlers(monitor, allItems);
+
+            numCrawlers = Math.min(allItems.size(), numCrawlers);
```

Comment:
could we avoid to change directly the class argument here and use local variable instead?

Reply:
sure thing